static/app/views/stories/storyTypes.tsx

```diff
@@ -0,0 +1,432 @@
+      // still search for it or expand its definitions
+      let expanded = true;
+      for (const node of Object.values(root.children)) {
+        node.visible = true;
+        node.expanded = expanded;
+        expanded = false;
+      }
+
+      return Object.values(root.children);
+    }
+
+    // Mark all as not matching
+    for (const {node} of root) {
+      node.visible = false;
+      node.expanded = false;
+      node.result = null;
+    }
+
+    // Fzf requires the input to be lowercase as it normalizes the search candidates to lowercase
+    const lowerCaseQuery = query.toLowerCase();
+
+    for (const {node, path} of root) {
+      // index files are useless when trying to match by name, so we'll special
+      // case them and match by their full path as it'll contain a more
+      // relevant path that we can match against.
+      const name = 'name' in node.prop ? node.prop.name : node.prop.definitionFilePath;
+      const match = fzf(name, lowerCaseQuery, false);
+
+      if (match.score > 0) {
+        node.result = match;
+        node.visible = true;
+
+        if (Object.keys(node.children).length > 0) {
+          node.expanded = true;
+          for (const child of Object.values(node.children)) {
+            child.visible = true;
+          }
+        }
+        for (const p of path) {
+          p.visible = true;
+          p.expanded = true;
+          // The entire path needs to contain max score of its child results so that
+          // the entire path to it can be sorted by this score. The side effect of this is that results from the same
+          // tree path with a lower score will be placed higher in the tree if that same path has a higher score anywhere
+          // in the tree. This isn't ideal, but given that it favors the most relevant results, it makes it a good starting point.
+          p.result = match.score > (p.result?.score ?? 0) ? match : p.result;
+        }
+      }
+    }
+
+    return Object.values(root.children);
+  }, [root, query]);
+
+  return nodes;
+}
+
+function propSort(a: PropTreeNode, b: PropTreeNode) {
+  if (a.visible && !b.visible) {
+    return -1;
+  }
+
+  if (!a.visible && b.visible) {
+    return 1;
+  }
+
+  if (a.result || b.result) {
+    if (a.result && b.result) {
+      return b.result.score - a.result.score;
+    }
+
+    if (a.result) {
+      return -1;
+    }
+
+    if (b.result) {
+      return 1;
+    }
+  }
+
+  const nameA = 'name' in a.prop ? a.prop.name : a.prop.definitionFilePath;
+  const nameB = 'name' in b.prop ? b.prop.name : b.prop.definitionFilePath;
+  return nameA.localeCompare(nameB);
+}
+
+function groupByParent(props: PropItem[]) {
+  return props.reduce(
+    (acc, prop) => {
+      acc[prop.parent?.fileName ?? ''] = (acc[prop.parent?.fileName ?? ''] ?? []).concat(
+        prop
+      );
+      return acc;
+    },
+    {} as Record<string, PropItem[]>
+  );
+}
```

Comment:
I think we could leverage the built-in `Object.groupBy` method here:

```
Object.groupBy(props, prop => prop.parent?.fileName ?? '');
```

Reply:
Til about this method